packages/@lwc/engine-core/src/framework/api.ts

```diff
@@ -686,10 +686,10 @@ export function fid(url: string | undefined | null): string | null | undefined {
     if (isNull(url)) {
         return null;
     }
-    const { renderMode } = vmBeingRendered;
+    const { idx, renderMode, shadowMode } = vmBeingRendered;
     // Apply transformation only for fragment-only-urls, and only in shadow DOM
-    if (/^#/.test(url) && renderMode === RenderMode.Shadow) {
-        return `${url}-${vmBeingRendered!.idx}`;
+    if (/^#/.test(url) && renderMode === RenderMode.Shadow && shadowMode === ShadowMode.Synthetic) {
```

Comment:
```suggestion
    if (renderMode === RenderMode.Shadow && shadowMode === ShadowMode.Synthetic && /^#/.test(url)) {
```

Micro-optimization, but we can put the integer equality checks before the regex test.